### DISCUSSION TITLE
questions about versioning, migrations and stop/resume use cases

### DISCUSSION THREAD

**Discussion by rullyalves** [cschleiden/go-workflows#357] [General]
Hello, first I would like to thank you for the great library, it offered me a great way to solve a problem that was causing me pain

I have some questions about implementation and specific use cases, could you help me?

1 - there are use cases where a workflow can be started, paused for a long time and then resume execution, if an activity calculates a volatile value, such as a user's account balance, is there any way to restart the workflow if it takes a long time to execute? I thought about using "continueAsNew" for this, but it would restart the entire workflow and not just specific activities
I thought about using continueAsNew to enable migrations without filling the code with if/else conditions, would it be viable to use continueAsNew for this? Or would it be better to keep previous versions of activities/workflows?

2 - about migrations, previously you answered that you could send the version as a parameter, but how do I know that all the workflows in the previous version have already been executed and that I can remove the old code without generating problems?

3 - I'm working with multiple instances, in case of migrations, what would be the best way to synchronize migrations between instances? Do I have any risks if I don't synchronize them?

4 - how can I get the value of a workflow to delete it? how to have a shopping cart status

I would like to try to provide solutions for these use cases or document patterns if they already exist

Finally, how could I contribute to the library? I thought the work was really magnificent and I would like to help with the features

**Comment by cschleiden** [comment]
> 1 - there are use cases where a workflow can be started, paused for a long time and then resume execution, if an activity calculates a volatile value, such as a user's account balance, is there any way to restart the workflow if it takes a long time to execute? I thought about using "continueAsNew" for this, but it would restart the entire workflow and not just specific activities I thought about using continueAsNew to enable migrations without filling the code with if/else conditions, would it be viable to use continueAsNew for this? Or would it be better to keep previous versions of activities/workflows?

I'm not sure I fully understand the use case. Activities should be relatively short-lived. Workflows can be active for a long time, but they can also schedule new activities to re-calculate values? 

> 2 - about migrations, previously you answered that you could send the version as a parameter, but how do I know that all the workflows in the previous version have already been executed and that I can remove the old code without generating problems?

There is really no support for this at the moment. As stated in the docs, the main scenario that I've targeted so far are side-by-side deployments where you deploy a new version. 

Using the new `Queues` feature to have separate queues per workflow version might work, however the stats endpoint doesn't expose the # of active instance per workflow queue yet, but should be pretty easy to query that yourself.

> 3 - I'm working with multiple instances, in case of migrations, what would be the best way to synchronize migrations between instances? Do I have any risks if I don't synchronize them?

There are risks, yes. Depending on the schema change you might have older workers accessing a newer schema. 

> 4 - how can I get the value of a workflow to delete it? how to have a shopping cart status

I'm not 100% following. Do you mean the workflow result for a workflow instance?
